lib/gc/Transforms/MemRefToCPURuntime.cpp

```diff
@@ -0,0 +1,156 @@
+    auto loc = op->getLoc();
+    MemRefType type = op.getMemref().getType();
+    ValueRange symbolOperands = op.getSymbolOperands();
+    ValueRange dynamicSizes = op.getDynamicSizes();
+    // Check if the parent contains a forall / parallel loop
+    bool hasParallelParent = false;
+    for (Operation *parentOp = op->getParentOp(); parentOp != nullptr;
+         parentOp = parentOp->getParentOp()) {
+      if (isa<scf::ForallOp, scf::ParallelOp>(parentOp)) {
+        hasParallelParent = true;
+        break;
+      }
+    }
+    Operation *newAllocOp;
+    if (hasParallelParent) {
+      newAllocOp = rewriter.create<cpuruntime::ThreadAllocOp>(
+          loc, type, dynamicSizes, symbolOperands);
+    } else {
+      newAllocOp = rewriter.create<cpuruntime::AllocOp>(loc, type, dynamicSizes,
+                                                        symbolOperands);
+    }
+    rewriter.replaceOp(op, newAllocOp->getResults());
+    return success();
+  }
+};
+
+struct AlignedDeallocLowering : public OpRewritePattern<memref::DeallocOp> {
+  using OpRewritePattern<memref::DeallocOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(memref::DeallocOp op,
+                                PatternRewriter &rewriter) const final {
+    auto loc = op->getLoc();
+    Value memref = op.getMemref();
+    // Check if the parent contains a forall / parallel loop
+    bool hasParallelParent = false;
+    for (Operation *parentOp = op->getParentOp(); parentOp != nullptr;
+         parentOp = parentOp->getParentOp()) {
+      if (isa<scf::ForallOp, scf::ParallelOp>(parentOp)) {
+        hasParallelParent = true;
+        break;
+      }
+    }
+    Operation *newDeallocOp;
+    if (hasParallelParent) {
+      newDeallocOp = rewriter.create<cpuruntime::ThreadDeallocOp>(loc, memref);
+    } else {
+      newDeallocOp = rewriter.create<cpuruntime::DeallocOp>(loc, memref);
+    }
+    rewriter.replaceOp(op, newDeallocOp->getResults());
+    return success();
+  }
+};
+
+struct ConvertMemRefToCPURuntime
+    : public impl::ConvertMemRefToCPURuntimeBase<ConvertMemRefToCPURuntime> {
+
+  void runOnOperation() final {
+    auto *ctx = &getContext();
+    // Create a local set to store operations that should not be transformed.
+    llvm::SmallSet<Operation *, 16> noTransformOps;
+
+    // Walk through the module to find func::FuncOp instances.
+    getOperation()->walk([&](func::FuncOp funcOp) {
+      BufferViewFlowAnalysis analysis(funcOp);
+      // Now walk through the operations within the func::FuncOp.
+      funcOp.walk([&](Operation *op) {
+        if (op->hasTrait<OpTrait::ReturnLike>()) {
+          for (Value operand : op->getOperands()) {
+            if (isa<MemRefType>(operand.getType())) {
+              SmallPtrSet<Value, 16> aliases = analysis.resolve(operand);
+              // Check if any of the returned memref is allocated within scope.
+              for (Value alias : aliases) {
```

Comment:
Suggest to use `const Value&` or `auto&&` to avoid copy.

Reply:
fixed.